### DISCUSSION TITLE
Using AWS IAM authentication with HikariCPJdbcDataSource

### DISCUSSION THREAD

**Discussion by alehuo** [slick/slick#2648] [Questions]
Hello!

Is there a way to use AWS IAM authentication with Slick's built-in Hikari connection pool? The database password needs to be refreshed every 15 minutes. It seems that Slick uses its own implementation of Hikari data source.

I've tried to use plugins that are found around GitHub and creating a new data source and configure Hikari to use it, with no success.

**Comment by nafg** [comment]
I don't know, how do you use it without Slick?

**Reply by alehuo** [reply]
Without Slick one would create a custom Hikari Data Source and configure it properly, so the connection timeout is e.g. `TOKEN_EXPIRATION - 1 minute`.

**Comment by alehuo** [comment]
Slick's Hikari connection pool code in question is here: https://github.com/slick/slick/blob/main/slick-hikaricp/src/main/scala/slick/jdbc/hikaricp/HikariCPJdbcDataSource.scala

This is how the flow currently goes:

1) `HikariDataSource` is created and the configuration is passed there (with a fixed password)
2) `HikariCPJdbcDataSource` is instantiated

The problem is that the password is set in the HikariDataSource and it apparently cannot be overwritten. I've followed this guide to implement password rotation, but with no success: https://blog.jdriven.com/2021/06/configure-hikari-connection-pool-when-using-rds-iam/

What I'd like to accomplish is to define a custom `HikariDataSource` instead of automatically creating it in the following line: https://github.com/slick/slick/blob/4ecde9a945a344184253ccd73de1cf9805baee36/slick-hikaricp/src/main/scala/slick/jdbc/hikaricp/HikariCPJdbcDataSource.scala#L93

Maybe it would be possible by using Scala's reflection, by defining the class name to use when instantiating the data source.

**Comment by rowi1de** [comment]
Are you able to use MariaDB Driver?
It works out of the box https://github.com/mariadb-corporation/mariadb-connector-j/blob/8acdd27379421149a25558eef8151defc84ed712/src/main/java/org/mariadb/jdbc/plugin/credential/aws/AwsIamCredentialPlugin.java#L43

just add 
```kotlin
  put("credentialType", "AWS-IAM")
```
